### PR TITLE
Csv related stuff (Work in progress)

### PR DESCRIPTION
**Pull request by snimavat**

### CODE REVIEW

File: gorm-tools-csv/src/main/groovy/gorm/tools/csv/CSVMapReader.groovy [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    //takes the first header line (assumed from firstreadline) from the file and sets the keys
+    def initFieldKeys() {
+        fieldKeys = fieldKeys ?: csvReader.readNext()
+    }
+
+    /**
+     * Reads the next line from the buffer in CSVReader and converts to map.
+     */
+    Map readNext() throws IOException {
+
```
Comment (jdabal) [inline gorm-tools-csv/src/main/groovy/gorm/tools/csv/CSVMapReader.groovy:75]: use PathKeyMap

Comment (snimavat) [reply gorm-tools-csv/src/main/groovy/gorm/tools/csv/CSVMapReader.groovy:75]: CsvPathMapkey reader is used now

File: gorm-tools/src/test/groovy/gorm/tools/databinding/EntityMapBinderUnitSpec.groovy [hunk 2]
```diff
@@ -481,6 +482,28 @@ class EntityMapBinderUnitSpec extends Specification implements DataRepoTest {
         TestEnumIdent.get(2) == TestEnumIdent.Num2
         testDomain.enumIdent == TestEnumIdent.Num2
     }
+
+    void "test bind using PathKeyMap"() {
```
Comment (basejump) [inline gorm-tools/src/test/groovy/gorm/tools/databinding/EntityMapBinderUnitSpec.groovy:486]: break into its own test

File: gorm-tools-csv/src/test/groovy/gorm/tools/csv/CSVReaderUtilsTest.groovy [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import com.opencsv.CSVReader
+
+/**
+ * Unit tests for the  {@link CSVReaderUtils}  class.
+ *
+ * @since 0.1
+ * @author Les Hazlewood
+ */
+
+class CSVReaderUtilsTest extends GroovyTestCase{
```
Comment (basejump) [inline gorm-tools-csv/src/test/groovy/gorm/tools/csv/CSVReaderUtilsTest.groovy:28]: remove this

File: gorm-tools-csv/src/test/groovy/gorm/tools/csv/CSVMapReaderTest.groovy [hunk 4]
```diff
@@ -0,0 +1,97 @@
+package gorm.tools.csv
+
+
+class CSVMapReaderTest extends GroovyTestCase {
```
Comment (basejump) [inline gorm-tools-csv/src/test/groovy/gorm/tools/csv/CSVMapReaderTest.groovy:4]: remove?

File: gorm-tools-csv/src/test/groovy/gorm/tools/csv/CSVMapReaderBatchTest.groovy [hunk 5]
```diff
@@ -0,0 +1,53 @@
+package gorm.tools.csv
+
+
+class CSVMapReaderBatchTest extends GroovyTestCase {
```
Comment (basejump) [inline gorm-tools-csv/src/test/groovy/gorm/tools/csv/CSVMapReaderBatchTest.groovy:4]: remove

File: gorm-tools-csv/src/main/groovy/gorm/tools/csv/CSVReaderUtils.groovy [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+import com.opencsv.ICSVParser
+import com.opencsv.validators.LineValidatorAggregator
+import com.opencsv.validators.RowValidatorAggregator
+
+/**
+ * Utility class for adding CSV parsing capability to core Java/Groovy classes (String, File, InputStream, Reader).
+ *
+ * @since 0.1* @author Les Hazlewood
+ */
+class CSVReaderUtils {
```
Comment (basejump) [inline gorm-tools-csv/src/main/groovy/gorm/tools/csv/CSVReaderUtils.groovy:18]: remove, is this helpful?

File: gorm-tools-csv/src/main/groovy/gorm/tools/csv/CSVPathKeyMapReader.groovy [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+package gorm.tools.csv
+
+import groovy.transform.CompileStatic
+import groovy.transform.builder.Builder
+import groovy.transform.builder.SimpleStrategy
+
+import com.opencsv.CSVReaderHeaderAware
+import gorm.tools.databinding.PathKeyMap
+
+@Builder(builderStrategy = SimpleStrategy, prefix = "")
```
Comment (basejump) [inline gorm-tools-csv/src/main/groovy/gorm/tools/csv/CSVPathKeyMapReader.groovy:14]: lets add simple javadoc, bread crumb on what this is for

File: gorm-tools-csv/src/main/groovy/gorm/tools/csv/CSVPathKeyMapReader.groovy [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        new CSVPathKeyMapReader(reader)
+    }
+
+    @Override
+    Map<String, String> readMap() {
+        Map<String, String> data = super.readMap()
+        return (Map<String, String>) new PathKeyMap(data, pathDelimiter)
+    }
+
+    List<Map<String, String>> readAllRows() {
```
Comment (basejump) [inline gorm-tools-csv/src/main/groovy/gorm/tools/csv/CSVPathKeyMapReader.groovy:48]: is this override? is there a super method that does this?

File: gorm-tools-csv/src/main/groovy/gorm/tools/csv/CSVMapReader.groovy [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+/**
+ * Class the read csv and return the rows as a map assuming the first row has the field/key names or you explicitly set fieldKeys property
+ * Also removes blank lines as for whatever reason the CSVReader does not
+ * Uses an underlying OpenCSV's CSVReader.
+ *
+ * @since 0.2* @author Joshua Burnett
+ */
+@CompileStatic
+class CSVMapReader implements Closeable, Iterable {
```
Comment (basejump) [inline gorm-tools-csv/src/main/groovy/gorm/tools/csv/CSVMapReader.groovy:19]: dod we need this? remove

File: gorm-tools-csv/grails-app/conf/application.yml [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+            check-template-location: false
+
+# Spring Actuator Endpoints are Disabled by Default
+endpoints:
+    enabled: false
+    jmx:
+        enabled: true
+---
+
+---
```
Comment (basejump) [inline gorm-tools-csv/grails-app/conf/application.yml:24]: skinny down to just what we need, see gorm-tools-security for example

File: gorm-tools-csv/grails-app/conf/application.groovy [hunk 11]
```diff
@@ -0,0 +1,11 @@
+grails {
```
Comment (basejump) [inline gorm-tools-csv/grails-app/conf/application.groovy:1]: this needed?

File: gorm-test-domain/src/main/groovy/yakworks/gorm/testing/model/KitchenSinkRepo.groovy [hunk 12]
```diff
@@ -48,6 +49,10 @@ class KitchenSinkRepo implements GormRepo<KitchenSink> {
         o
     }
 
+    void doAssociations(KitchenSink kitchenSink, Map data) {
```
Comment (basejump) [inline gorm-test-domain/src/main/groovy/yakworks/gorm/testing/model/KitchenSinkRepo.groovy:52]: this is not called, do we not need it?

File: gorm-tools-csv/src/test/groovy/gorm/tools/csv/CSVPathKeyMapReaderSpec.groovy [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    void "merge two csv into single map structure"() {
+        setup:
+        CSVPathKeyMapReader kitchenSinkReader = new CSVPathKeyMapReader(new FileReader(kitchenSinkCsv))
+        CSVPathKeyMapReader sinkItemReader = new CSVPathKeyMapReader(new FileReader(sinkItemCsv), "_")
+
+        when:
+        List<Map> kitchenSinks = kitchenSinkReader.readAllRows()
+        List<Map> sinkItems = []
+        while(sinkItemReader.hasNext()) {
+            Map row = sinkItemReader.readMap() { PathKeyMap m ->
```
Comment (basejump) [inline gorm-tools-csv/src/test/groovy/gorm/tools/csv/CSVPathKeyMapReaderSpec.groovy:72]: are we not going to use closure? add comments/bread crumbs on whats going on here

File: examples/benchmarks/src/main/groovy/gpbench/helpers/csv/CSVMapReader.groovy [hunk 14]
```diff
@@ -26,7 +27,7 @@ import au.com.bytecode.opencsv.CSVReader
  * @since 0.2
  * @author Joshua Burnett
  */
-public class CSVMapReader implements Closeable, Iterable {
+class CSVMapReader implements Closeable, Iterable {
```
Comment (basejump) [inline examples/benchmarks/src/main/groovy/gpbench/helpers/csv/CSVMapReader.groovy:30]: remove if we dont need it, which I dont see why we do.

**Review by basejump — CHANGES_REQUESTED** [review]
see review

File: gorm-tools-csv/src/main/groovy/gorm/tools/csv/CSVPathKeyMapReader.groovy [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+        Map<String, String> data = super.readMap()
+        return (Map<String, String>) new PathKeyMap(data, pathDelimiter)
+    }
+
+    /**
+     * Read all the rows in CSV, we can't override the readAll in CVSReader as is return list of string array
+     *
+     * @return the list of maps for entire file
+     */
+    List<Map<String, String>> readAllRows() {
```
Comment (jdabal) [inline gorm-tools-csv/src/main/groovy/gorm/tools/csv/CSVPathKeyMapReader.groovy:59]: rename readAllRows to readAllLines

**Pull request merged**